FeedStoreChallenge/RealmFeedStore.swift

```
@@ -0,0 +1,163 @@
+		guard let uuid = UUID(uuidString: id),
+			  let url = URL(string: url) else {
+			 return nil
+		}
+		return LocalFeedImage(id: uuid, description: desc, location: location, url: url)
+	}
+}
+
+final class RealmFeedCache: Object {
+	
+	@objc fileprivate dynamic var timestamp: Date
+	fileprivate dynamic var images = List<RealmFeedStoreCachedFeedImage>()
+	
+	fileprivate override init() {
+		timestamp = Date.distantPast
+		super.init()
+	}
+	
+	convenience init(timestamp: Date, feed: [LocalFeedImage]) {
+		self.init()
+		self.timestamp = timestamp
+		self.images.append(objectsIn: feed.map(RealmFeedStoreCachedFeedImage.init(localFeedImage:)))
+	}
+}
+
+// MARK:-
+public final class RealmFeedStore: FeedStore {
+	
+	private let configuration: Realm.Configuration
+	private let queue: DispatchQueue
+	
+	enum WriteError: Error { case readonlyStore }
+
+	public enum Configuration {
+		case fileURL(URL)
+		case inMemoryIdentifier(String)
+	}
+	
+	/// use this initializer for testing purposes only
+	///
+	/// use init(fileURL: URL) for production code
+	public init(_ configuration:Configuration, readOnly: Bool) {
+		switch configuration {
+		case .fileURL(let URL):
+		self.configuration = Realm.Configuration(fileURL: URL, readOnly: readOnly)
+		case .inMemoryIdentifier(let identifier):
+		self.configuration = Realm.Configuration(inMemoryIdentifier: identifier, readOnly: readOnly)
+		}
+		self.queue = DispatchQueue(label: "\(type(of: Self.self))", qos: .userInitiated, autoreleaseFrequency: .workItem)
+	}
+	
+	/// use this initializer for production code
+	public convenience init(fileURL: URL) {
+		self.init(.fileURL(fileURL), readOnly: false)
+	}
+	
+	private var _realm: Realm?
+	private func getRealm() throws -> Realm {
+		if let existing = _realm { return existing }
+		_realm = try Realm(configuration: self.configuration, queue: queue)
+		return _realm!
+	}
+	
+	private var canWrite: Bool { !configuration.readOnly }
+	private func ensureSafeConfiguration() -> Error? {
```

It's more idiomatic to throw errors instead of returning optional errors in Swift.

---

I considered this.  Throwing from this method would require either 2 do/catch blocks everywhere it's called (one for calling this method and one for the actual work, as it happens on a different queue), or entering a separate queue just to then fail this test.

I consider this to be more or less a precondition at the call site:
`		if let error = ensureSafeConfiguration() { return completion(error) }
`
It it can't work, then throw right away.

It's less idiomatic, but I think it fits this use case better.

---

Yes, the idea was to have a single do/catch where you can call the `ensureSafeConfiguration`. But that would require running it inside the queue block.

The problem is that this class allows an invalid configuration (readonly) - The interface says you can insert/delete, but internally it might refuse the operation because of the readonly state. This class shouldn't allow read-only configurations if it cannot perform all its operations with it.

The readonly config helps with tests but complicates the implementation. it's a trade-off as always 🙂

---

with the new init structure, the readonly option is only available to test code.  Production code can't access it. So hopefully the tradeofff leans toward the positive...

---

Looks good!